### PR TITLE
wasmbinding test

### PR DESCRIPTION
**Pull request by expertdicer**

### CODE REVIEW

File: wasmbinding/test/helpers_test.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+)
+
+func CreateTestInput(t *testing.T) (*app.App, sdk.Context) {
+	junoApp := app.Setup(t, false, 1)
+	ctx := junoApp.BaseApp.NewContext(false, tmtypes.Header{Height: 1, ChainID: "kujira-1", Time: time.Now().UTC()})
+	return junoApp, ctx
+}
+
+// we need to make this deterministic (same every test run), as content might affect gas costs
+func keyPubAddr() (crypto.PrivKey, crypto.PubKey, sdk.AccAddress) {
```
Comment (catShaark) [inline wasmbinding/test/helpers_test.go:23]: are these functions used ?

File: wasmbinding/test/custom_query_test.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+		ExchangeRate: &wasm.ExchangeRateQueryParams{
+			Denom: "b",
+		},
+	}
+	resp := oracle.ExchangeRateQueryResponse{}
+	queryRate(t, ctx, junoApp, oracleQuerier, query, &resp)
+
+}
+
+func queryRate(t *testing.T, ctx sdk.Context, junoApp *app.App, contract sdk.AccAddress, request oracle.OracleQuery, response interface{}) {
```
Comment (catShaark) [inline wasmbinding/test/custom_query_test.go:62]: This shouldn't be a function on its own, also you should compare the contract state with a query to oraclekeeper instead a hardcode "1.7"

File: wasmbinding/test/custom_query_test.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+	oracle "github.com/CosmosContracts/juno/v11/x/oracle/wasm"
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/stretchr/testify/require"
+)
+
+func SetupCustomApp(t *testing.T, addr sdk.AccAddress) (*app.App, sdk.Context) {
+	junoApp, ctx := CreateTestInput(t)
+	wasmKeeper := junoApp.WasmKeeper()
+	storeOracleQuerierCode(t, ctx, junoApp, addr)
```
Comment (catShaark) [inline wasmbinding/test/custom_query_test.go:22]: This shouldn't be a function on its own

**Review by catShaark — CHANGES_REQUESTED** [review]

File: wasmbinding/test/custom_query_test.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+	actorAmount := sdk.NewCoins(sdk.NewCoin(appparams.BondDenom, sdk.NewInt(100000000000000)))
+	fundAccount(t, ctx, junoApp, actor, actorAmount)
+
+	ExchangeRateC := sdk.NewDec(1700)
+	ExchangeRateB := sdk.NewDecWithPrec(17, 1)
+	ExchangeRateD := sdk.NewDecWithPrec(19, 1)
+	junoApp.OracleKeeper.SetExchangeRate(ctx, "a", sdk.NewDec(1))
+	junoApp.OracleKeeper.SetExchangeRate(ctx, "b", ExchangeRateB)
+	junoApp.OracleKeeper.SetExchangeRate(ctx, "c", ExchangeRateC)
```
Comment (catShaark) [inline wasmbinding/test/custom_query_test.go:45]: why setting all the price for denom a, b, c when only use denom b

File: wasmbinding/test/custom_query_test.go [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+	err := simapp.FundAccount(
+		juno.BankKeeper,
+		ctx,
+		addr,
+		coins,
+	)
+	require.NoError(t, err)
+}
+
+func executeRawCustom(t *testing.T, ctx sdk.Context, juno *app.App, contract sdk.AccAddress, sender sdk.AccAddress, msg json.RawMessage, funds sdk.Coin) error {
```
Comment (catShaark) [inline wasmbinding/test/custom_query_test.go:137]: this should also be in app folder

**Review by catShaark — CHANGES_REQUESTED** [review]

File: wasmbinding/test/custom_query_test.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+	initMsgBz := []byte("{}")
+	contractKeeper := keeper.NewDefaultPermissionKeeper(junoApp.WasmKeeper())
+	codeID := uint64(1)
+	addr, _, err := contractKeeper.Instantiate(ctx, codeID, funder, funder, initMsgBz, "demo contract", nil)
+	require.NoError(t, err)
+
+	return addr
+}
+
+func fundAccount(t *testing.T, ctx sdk.Context, juno *app.App, addr sdk.AccAddress, coins sdk.Coins) {
```
[the remark this answers is not in the document]
Comment (expertdicer) [reply wasmbinding/test/custom_query_test.go:127]: Removed

File: wasmbinding/test/helpers_test.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package wasmbinding_test
+
+import (
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/ed25519"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func keyPubAddr() (crypto.PrivKey, crypto.PubKey, sdk.AccAddress) {
```
[the remark this answers is not in the document]
Comment (expertdicer) [reply wasmbinding/test/helpers_test.go:10]: Done

File: wasmbinding/test/custom_query_test.go [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+	require.NoError(t, err)
+
+	resBz, err := junoApp.WasmKeeper().QuerySmart(ctx, contract, queryBz)
+	var rate string
+	json.Unmarshal(resBz, &rate)
+	require.NoError(t, err)
+	require.Equal(t, rate, "1.7")
+}
+
+func storeOracleQuerierCode(t *testing.T, ctx sdk.Context, junoApp *app.App, addr sdk.AccAddress) {
```
[the remark this answers is not in the document]
Comment (expertdicer) [reply wasmbinding/test/custom_query_test.go]: Done, maybe

**Pull request merged**